### PR TITLE
generate doc for latest published crate

### PR DESCRIPTION
**Pull request by akash-fortanix**
Fix for https://github.com/fortanix/rust-sgx/issues/130

Fix breaks currently since enclave-runner's metadata specifies version 0.1.1 but the latest published version is 0.1.0

### CODE REVIEW

File: doc/generate-api-docs.sh [hunk 1]
```diff
@@ -49,12 +49,30 @@ LIBS_SORTED=$(
...
+dependencies=''
+
 for LIB in $LIBS_SORTED; do
     cd $LIB
+    version=$(cat Cargo.toml | grep version | head -n1 | cut -d'"' -f2)
+    dependency=$LIB' = { version = "'$version'"'
     ARGS=""
     if FEATURES="$(cargo read-manifest|jq -r '.metadata.docs.rs.features | join(",")' 2> /dev/null)"; then
         ARGS="--features $FEATURES"
+        dependency=$dependency', features = ["'$FEATURES'"]'
```
Comment (akash-fortanix) [inline doc/generate-api-docs.sh]: Incorrect template for multiple features

File: doc/generate-api-docs.sh [hunk 2]
```diff
@@ -49,12 +49,30 @@ LIBS_SORTED=$(
     | egrep '^'$(echo $LIBS|sed 's/ /|/g')'$'
 )
 
+dependencies=''
+
 for LIB in $LIBS_SORTED; do
     cd $LIB
+    version=$(cat Cargo.toml | grep version | head -n1 | cut -d'"' -f2)
```
Comment (akash-fortanix) [inline doc/generate-api-docs.sh]: As discussed, getting version from latest git tag should be the way, this will resolve the issue with this fix around enclave-runner.

File: doc/generate-api-docs.sh [hunk 3]
```diff
@@ -49,12 +49,30 @@ LIBS_SORTED=$(
...
+    dependencies=$dependencies$dependency$'\n'
+    #cargo doc --no-deps --lib $ARGS
     cd -
 done
+
+pushd $WORKSPACE
+cargo new foo
+cd foo
+echo "$dependencies" >> Cargo.toml
+cargo doc
```
Comment (jethrogb) [inline doc/generate-api-docs.sh]: Doesn't this generate documentation for crate foo as well?

Comment (akash-fortanix) [reply doc/generate-api-docs.sh]: It does, but I copy the libs' docs to rust-sgx/target/doc

Comment (akash-fortanix) [reply doc/generate-api-docs.sh]: The cargo doc works for:

sgx-isa = { version = "0.2.0" }
rs-libc = { version = "0.1.0" }
fortanix-sgx-abi = { version = "0.3.2", features = ["docs"] }
dcap-ql-sys = { version = "0.1.0" }
sgxs = { version = "0.6.0" }
sgxs-loaders = { version = "0.1.0" }

but fails for 

enclave-runner = { version = "0.1.0" }

and onwards

report-test = { version = "0.1.0" }
dcap-ql = { version = "0.1.0" }
aesm-client = { version = "0.1.0" }

I need to resolve this issue.

File: doc/generate-api-docs.sh [hunk 4]
```diff
@@ -51,10 +51,29 @@ LIBS_SORTED=$(
...
     fi
-    cargo doc --no-deps --lib $ARGS
+    dependency=$dependency' }'
     cd -
+
+    pushd $WORKSPACE
+    cargo new foo
+    cd foo
+    echo "$dependency" >> Cargo.toml
+    cargo doc -p $LIB
```
Comment (jethrogb) [inline doc/generate-api-docs.sh]: Shouldn't this be run with `--no-deps`?

File: doc/generate-api-docs.sh [hunk 5]
```diff
@@ -51,10 +51,29 @@ LIBS_SORTED=$(
...
+    popd
+
+    ls -lrta $WORKSPACE/foo/target/doc/
+
+    _LIB=${LIB//-/_}
+
+    cp -r $WORKSPACE/foo/target/doc/$_LIB target/doc
+
+    pushd $WORKSPACE
+    rm -rf foo
```
Comment (jethrogb) [inline doc/generate-api-docs.sh]: I think it would be more efficient to use a single crate `foo` that is used to generate all the docs instead of creating a new workspace every time.

File: doc/generate-api-docs.sh [hunk 6]
```diff
@@ -51,10 +51,29 @@ LIBS_SORTED=$(
...
+    dependency=$LIB' = { version = "'$version'"'
+    features="$(cargo read-manifest|jq -c '.metadata.docs.rs.features')"
+    if [ $features != 'null' ]; then
+        dependency=$dependency', features = '$features
     fi
-    cargo doc --no-deps --lib $ARGS
+    dependency=$dependency' }'
     cd -
+
+    pushd $WORKSPACE
```
Comment (jethrogb) [inline doc/generate-api-docs.sh]: Where is this set?

File: doc/generate-api-docs.sh [hunk 7]
```diff
@@ -51,10 +51,29 @@ LIBS_SORTED=$(
 
 for LIB in $LIBS_SORTED; do
     cd $LIB
-    ARGS=""
-    if FEATURES="$(cargo read-manifest|jq -r '.metadata.docs.rs.features | join(",")' 2> /dev/null)"; then
-        ARGS="--features $FEATURES"
+    version=$(git tag --sort=taggerdate | grep $LIB'_' | tail -n1 | cut -d'_' -f2 | cut -d'v' -f2)
+    dependency=$LIB' = { version = "'$version'"'
```
Comment (jethrogb) [inline doc/generate-api-docs.sh]: You'll want to use an exact version specification: `version = "='$version'"`

File: doc/generate-api-docs.sh (hunk 5 above)
Comment (akash-fortanix) [reply doc/generate-api-docs.sh]: That was the initial approach, but it didn't work out > https://github.com/fortanix/rust-sgx/pull/131#discussion_r276539677

File: doc/generate-api-docs.sh (hunk 6 above)
Comment (akash-fortanix) [reply doc/generate-api-docs.sh]: It's an environment variable set for this process, should I check if its set before its used?

File: doc/generate-api-docs.sh (hunk 4 above)
Comment (akash-fortanix) [reply doc/generate-api-docs.sh]: Yes, done.

File: doc/generate-api-docs.sh (hunk 7 above)
Comment (akash-fortanix) [reply doc/generate-api-docs.sh]: Done.

File: doc/generate-api-docs.sh (hunk 6 above)
Comment (jethrogb) [reply doc/generate-api-docs.sh]: Yes

File: doc/generate-api-docs.sh (hunk 5 above)
Comment (jethrogb) [reply doc/generate-api-docs.sh]: Too bad

Comment (jethrogb) [reply doc/generate-api-docs.sh]: That comment doesn't mention what the issue was you were having.

Thinking more about it I don't think the current approach is not just inefficient, it's also incorrect. We want to fix the version of all of our crates that are being used in the doc generation, including those used as dependencies of the crate currently being documented.

Comment (akash-fortanix) [reply doc/generate-api-docs.sh]: Updating crates.io index
   Compiling proc-macro2 v0.4.29
   Compiling unicode-xid v0.1.0
   Compiling semver-parser v0.7.0
   Compiling syn v0.15.33
   Compiling libc v0.2.53
   Compiling cc v1.0.35
   Compiling pkg-config v0.3.14
   Compiling autocfg v0.1.2
   Compiling failure_derive v0.1.5
    Checking rustc-demangle v0.1.14
   Compiling byteorder v1.3.1
    Checking foreign-types-shared v0.1.1
   Compiling openssl v0.10.20
    Checking cfg-if v0.1.7
    Checking lazy_static v1.3.0
    Checking bitflags v1.0.4
   Compiling nix v0.13.0
    Checking void v1.0.2
    Checking fnv v1.0.6
    Checking fortanix-sgx-abi v0.3.2
   Compiling semver v0.9.0
   Compiling backtrace v0.3.15
    Checking foreign-types v0.3.2
    Checking sgx-isa v0.2.0
   Compiling backtrace-sys v0.1.28
   Compiling rustc_version v0.2.3
    Checking time v0.1.42
   Compiling openssl-sys v0.9.43
   Compiling quote v0.6.12
   Compiling synstructure v0.10.1
    Checking failure v0.1.5
    Checking sgxs v0.6.0
 Documenting enclave-runner v0.1.0
error[E0432]: unresolved import `self::abi::dispatch`
  --> /home/akash/.cargo/registry/src/github.com-1ecc6299db9ec823/enclave-runner-0.1.0/src/usercalls/mod.rs:37:5
   |
37 | use self::abi::dispatch;
   |     ^^^^^^^^^^^^^^^^^^^ no `dispatch` in `usercalls::abi`

error: cannot find macro `invoke_with_usercalls!` in this scope
   --> /home/akash/.cargo/registry/src/github.com-1ecc6299db9ec823/enclave-runner-0.1.0/src/usercalls/abi.rs:299:1
    |
299 | invoke_with_usercalls!(define_usercalls);
    | ^^^^^^^^^^^^^^^^^^^^^

error[E0404]: expected trait, found struct `Usercalls`
  --> /home/akash/.cargo/registry/src/github.com-1ecc6299db9ec823/enclave-runner-0.1.0/src/usercalls/interface.rs:19:10
   |
19 | impl<'a> Usercalls for Handler<'a> {
   |          ^^^^^^^^^ not a trait

error[E0412]: cannot find type `FifoDescriptor` in this scope
   --> /home/akash/.cargo/registry/src/github.com-1ecc6299db9ec823/enclave-runner-0.1.0/src/usercalls/interface.rs:144:30
    |
144 |         usercall_queue: *mut FifoDescriptor<Usercall>,
    |                              ^^^^^^^^^^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
    |
9   | use fortanix_sgx_abi::async::FifoDescriptor;
    |

error[E0412]: cannot find type `Usercall` in this scope
   --> /home/akash/.cargo/registry/src/github.com-1ecc6299db9ec823/enclave-runner-0.1.0/src/usercalls/interface.rs:144:45
    |
144 |         usercall_queue: *mut FifoDescriptor<Usercall>,
    |                                             ^^^^^^^^
help: a struct with a similar name exists
    |
144 |         usercall_queue: *mut FifoDescriptor<Usercalls>,
    |                                             ^^^^^^^^^
help: possible candidates are found in other modules, you can import them into scope
    |
9   | use fortanix_sgx_abi::async::Usercall;
    |
9   | use tcs::Usercall;
    |

error[E0412]: cannot find type `FifoDescriptor` in this scope
   --> /home/akash/.cargo/registry/src/github.com-1ecc6299db9ec823/enclave-runner-0.1.0/src/usercalls/interface.rs:145:28
    |
145 |         return_queue: *mut FifoDescriptor<Return>,
    |                            ^^^^^^^^^^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
    |
9   | use fortanix_sgx_abi::async::FifoDescriptor;
    |

error[E0412]: cannot find type `Return` in this scope
   --> /home/akash/.cargo/registry/src/github.com-1ecc6299db9ec823/enclave-runner-0.1.0/src/usercalls/interface.rs:145:43
    |
145 |         return_queue: *mut FifoDescriptor<Return>,
    |                                           ^^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
    |
9   | use fortanix_sgx_abi::async::Return;
    |

error[E0412]: cannot find type `FifoDescriptor` in this scope
   --> /home/akash/.cargo/registry/src/github.com-1ecc6299db9ec823/enclave-runner-0.1.0/src/usercalls/mod.rs:903:30
    |
903 |         usercall_queue: &mut FifoDescriptor<Usercall>,
    |                              ^^^^^^^^^^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
    |
10  | use fortanix_sgx_abi::async::FifoDescriptor;
    |

error[E0412]: cannot find type `Usercall` in this scope
   --> /home/akash/.cargo/registry/src/github.com-1ecc6299db9ec823/enclave-runner-0.1.0/src/usercalls/mod.rs:903:45
    |
903 |         usercall_queue: &mut FifoDescriptor<Usercall>,
    |                                             ^^^^^^^^
help: a struct with a similar name exists
    |
903 |         usercall_queue: &mut FifoDescriptor<Usercalls>,
    |                                             ^^^^^^^^^
help: possible candidates are found in other modules, you can import them into scope
    |
10  | use fortanix_sgx_abi::async::Usercall;
    |
10  | use tcs::Usercall;
    |

error[E0412]: cannot find type `FifoDescriptor` in this scope
   --> /home/akash/.cargo/registry/src/github.com-1ecc6299db9ec823/enclave-runner-0.1.0/src/usercalls/mod.rs:904:28
    |
904 |         return_queue: &mut FifoDescriptor<Return>,
    |                            ^^^^^^^^^^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
    |
10  | use fortanix_sgx_abi::async::FifoDescriptor;
    |

error[E0412]: cannot find type `Return` in this scope
   --> /home/akash/.cargo/registry/src/github.com-1ecc6299db9ec823/enclave-runner-0.1.0/src/usercalls/mod.rs:904:43
    |
904 |         return_queue: &mut FifoDescriptor<Return>,
    |                                           ^^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
    |
10  | use fortanix_sgx_abi::async::Return;
    |

error[E0603]: struct `Usercalls` is private
  --> /home/akash/.cargo/registry/src/github.com-1ecc6299db9ec823/enclave-runner-0.1.0/src/usercalls/interface.rs:14:34
   |
14 | use super::abi::{UsercallResult, Usercalls};

Comment (akash-fortanix) [reply doc/generate-api-docs.sh]: foo's Cargo.toml

[dependencies]
sgx-isa = { version = "=0.2.0" }
rs-libc = { version = "=0.1.0" }
fortanix-sgx-abi = { version = "=0.3.2", features = ["docs"]}
sgxs = { version = "=0.6.0" }
dcap-ql-sys = { version = "=0.1.0" }
sgxs-loaders = { version = "=0.1.0" }
enclave-runner = { version = "=0.1.0" }
report-test = { version = "=0.1.0" }
dcap-ql = { version = "=0.1.0" }
aesm-client = { version = "=0.1.0" }

and

cargo doc -p enclave-runner --no-deps

causes above.

Basically, features = ["docs"]

fortanix-sgx-abi = { version = "=0.3.2", features = ["docs"]}

causes failure.

I'm able to generate docs by omitting docs feature from foo's Cargo.toml

Comment (akash-fortanix) [reply doc/generate-api-docs.sh]: Have updated the PR incorporating above, but it breaks the way fortanix-sgx-abi doc was generated earlier.

File: doc/generate-api-docs.sh (hunk 6 above)
Comment (akash-fortanix) [reply doc/generate-api-docs.sh]: Done

File: doc/generate-api-docs.sh (hunk 5 above)
Comment (jethrogb) [reply doc/generate-api-docs.sh]: Ah I understand now. I'm not sure yet what's the best way to handle this. Fundamentally, generating the docs for `fortanix_sgx_abi` is incompatible with using that crate as a dependency.

Comment (jethrogb) [reply doc/generate-api-docs.sh]: I think we can make an exception in how we generate the docs for `fortanix_sgx_abi`. We prepare the Cargo.toml as you do now, but don't generate docs for `fortanix_sgx_abi`. Then, generate docs separately for `fortanix_sgx_abi` with the docs feature. This is ok because the `fortanix_sgx_abi` types are not used in any public API.

**Comment by parthsane** [comment]
@akash-fortanix could you rebase this branch?
